doc/api/async_hooks.md

```diff
@@ -0,0 +1,670 @@
+  needs to be released during _destroy_
+
+Called when a class is constructed that has the _possibility_ to trigger an
+asynchronous event. This _does not_ mean the instance must trigger
+`before()`/`after()` before `destroy()` is called. Only that the possibility
+exists.
+
+This behavior can be observed by doing something like opening a resource then
+closing it before the resource can be used. The following snippet demonstrates
+this.
+
+```js
+require('net').createServer().listen(function() { this.close() });
+// OR
+clearTimeout(setTimeout(() => {}, 10));
+```
+
+Every new resource is assigned a unique id.
+
+The `type` is a String that represents the type of resource that caused
+`init()` to fire. Generally it will be the name of the resource's constructor.
+Some examples include `TCP`, `GetAddrInfo` and `HTTPParser`. Users will be able
+to define their own `type` when using the public embedder API.
+
+**Note:** It is possible to have type name collisions. Embedders are recommended
+to use unique prefixes per module to prevent collisions when listening to the
+hooks.
+
+`triggerId` is the `id` of the resource that caused (or "triggered") the
+new resource to initialize and that caused `init()` to fire.
+
+The following is a simple demonstration of this:
+
+```js
+const async_hooks = require('async_hooks');
+
+asyns_hooks.createHook({
+  init (id, type, triggerId) {
+    const cId = async_hooks.currentId();
+    process._rawDebug(`${type}(${id}): trigger: ${triggerId} scope: ${cId}`);
+  }
+}).enable();
+
+require('net').createServer(c => {}).listen(8080);
+```
+
+Output when hitting the server with `nc localhost 8080`:
+
+```
+TCPWRAP(2): trigger: 1 scope: 1
+TCPWRAP(4): trigger: 2 scope: 0
+```
+
+The second `TCPWRAP` is the new connection from the client. When a new
+connection is made the `TCPWrap` instance is immediately constructed. This
+happens outside of any JavaScript stack (side note: a `currentId()` of `0`
+means it's being executed in "the void", with no JavaScript stack above it).
+With only that information it would be impossible to link resources together in
+terms of what caused them to be created. So `triggerId` is given the task of
+propagating what resource is responsible for the new resource's existence.
+
+Below is another example with additional information about the calls to
+`init()` between the `before()` and `after()` calls. Specifically what the
+callback to `listen()` will look like. The output formatting is slightly more
+elaborate to make calling context easier to see.
+
+```js
+const async_hooks = require('async_hooks');
+
+let ws = 0;
+async_hooks.createHook({
+  init (id, type, triggerId) {
+    const cId = async_hooks.currentId();
+    process._rawDebug(' '.repeat(ws) +
+                      `${type}(${id}): trigger: ${triggerId} scope: ${cId}`);
+  },
+  before (id) {
+    process._rawDebug(' '.repeat(ws) + 'before: ', id);
```

Comment:
There is the option of not allowing any async hooks to execute while one of its own hooks is running. That should be the most straight forward approach, but we'd also loose information about what's happening during hook execution. If that's a fine trade then I can whip up the patch.

FYI: Reason `process._rawDebug()` needs to be used is because it writes directly to `stderr`. A similar effect would be to do `fs.writeSync(2, 'foo');`. Otherwise `init` will be called and cause a circular call to the hooks. Causing the application to crash.